client/csi_endpoint.go

```diff
@@ -311,6 +311,135 @@ func (c *CSI) ControllerListVolumes(req *structs.ClientCSIControllerListVolumesR
+	defer metrics.MeasureSince([]string{"client", "csi_controller", "create_snapshot"}, time.Now())
+
+	plugin, err := c.findControllerPlugin(req.PluginID)
+	if err != nil {
+		// the server's view of the plugin health is stale, so let it know it
+		// should retry with another controller instance
+		return fmt.Errorf("CSI.ControllerCreateSnapshot: %w: %v",
+			nstructs.ErrCSIClientRPCRetryable, err)
+	}
+	defer plugin.Close()
+
+	csiReq, err := req.ToCSIRequest()
+	if err != nil {
+		return fmt.Errorf("CSI.ControllerCreateSnapshot: %v", err)
+	}
+
+	ctx, cancelFn := c.requestContext()
+	defer cancelFn()
+
+	// CSI ControllerCreateSnapshot errors for timeout, codes.Unavailable and
+	// codes.ResourceExhausted are retried; all other errors are fatal.
+	cresp, err := plugin.ControllerCreateSnapshot(ctx, csiReq,
+		grpc_retry.WithPerRetryTimeout(CSIPluginRequestTimeout),
+		grpc_retry.WithMax(3),
+		grpc_retry.WithBackoff(grpc_retry.BackoffExponential(100*time.Millisecond)))
+	if err != nil {
+		return fmt.Errorf("CSI.ControllerCreateSnapshot: %v", err)
+	}
+
+	if cresp == nil || cresp.Snapshot == nil {
+		c.c.logger.Warn("plugin did not return error or snapshot; this is a bug in the plugin and should be reported to the plugin author")
+		return fmt.Errorf("CSI.ControllerCreateSnapshot: plugin did not return error or snapshot")
+	}
+	resp.ID = cresp.Snapshot.ID
+	resp.ExternalSourceVolumeID = cresp.Snapshot.SourceVolumeID
+	resp.SizeBytes = cresp.Snapshot.SizeBytes
+	resp.CreateTime = cresp.Snapshot.CreateTime
+	resp.IsReady = cresp.Snapshot.IsReady
+
+	return nil
+}
+
+func (c *CSI) ControllerDeleteSnapshot(req *structs.ClientCSIControllerDeleteSnapshotRequest, resp *structs.ClientCSIControllerDeleteSnapshotResponse) error {
+	defer metrics.MeasureSince([]string{"client", "csi_controller", "delete_snapshot"}, time.Now())
+
+	plugin, err := c.findControllerPlugin(req.PluginID)
+	if err != nil {
+		// the server's view of the plugin health is stale, so let it know it
+		// should retry with another controller instance
+		return fmt.Errorf("CSI.ControllerDeleteSnapshot: %w: %v",
+			nstructs.ErrCSIClientRPCRetryable, err)
+	}
+	defer plugin.Close()
+
+	csiReq := req.ToCSIRequest()
+
+	ctx, cancelFn := c.requestContext()
+	defer cancelFn()
+
+	// CSI ControllerDeleteSnapshot errors for timeout, codes.Unavailable and
+	// codes.ResourceExhausted are retried; all other errors are fatal.
+	err = plugin.ControllerDeleteSnapshot(ctx, csiReq,
+		grpc_retry.WithPerRetryTimeout(CSIPluginRequestTimeout),
+		grpc_retry.WithMax(3),
+		grpc_retry.WithBackoff(grpc_retry.BackoffExponential(100*time.Millisecond)))
+	if errors.Is(err, nstructs.ErrCSIClientRPCIgnorable) {
+		// if the snapshot was deleted out-of-band, we'll get an error from
+		// the plugin but can safely ignore it
+		c.c.logger.Debug("could not delete snapshot: %v", err)
```

Comment:
I'm not sure why `hclogvet` isn't flagging this :( 
```suggestion
		c.c.logger.Debug("could not delete snapshot", "error", err)
```

Reply:
Yeah, that's weird. Looks like there's a couple of these.

Reply:
Fixed in 3a19cd5